turbo/snapshotsync/freezeblocks/caplin_snapshots.go

```diff
@@ -280,21 +282,65 @@ Loop:
 		s.segmentsMax.Store(segmentsMax)
 	}
 	s.segmentsReady.Store(true)
+
+	s.calculateVisibleSegments()
 	s.idxMax.Store(s.idxAvailability())
 	s.indicesReady.Store(true)
 
 	return nil
 }
 
-func (s *CaplinSnapshots) idxAvailability() uint64 {
-	var beaconBlocks uint64
-	for _, seg := range s.BeaconBlocks.segments {
-		if seg.Index() == nil {
+func (s *CaplinSnapshots) calculateVisibleSegments() {
+	var maxIndex = -1
+	var maxBlockNum uint64
+	if len(s.BeaconBlocks.dirtySegments) == 0 {
+		return
+	}
+	slices.SortFunc(s.BeaconBlocks.dirtySegments, func(i, j *Segment) int {
+		return cmp.Compare(i.from, j.from)
+	})
+
+	if s.BeaconBlocks.maxVisibleBlock.Load() != 0 &&
+		s.BeaconBlocks.maxVisibleBlock.Load()+1 != s.BeaconBlocks.dirtySegments[0].from {
+		s.logger.Warn("gap in beacon blocks snapshots", "from", s.BeaconBlocks.maxVisibleBlock.Load(), "to", s.BeaconBlocks.dirtySegments[0].from)
```

Comment:
It’s unclear - why this warning is showed “before calculation of visible files is done

Reply:
Because we're able to detect the gap before appending since both visible and dirty segments are sorted by from/to/version.